docs/source/getting-started/testing.rst

```diff
@@ -0,0 +1,105 @@
+We use it to spawn the ``daml start`` and ``yarn start`` processes and launch the browser.
+On the other hand the ``afterAll()`` section is used to shut down these processes and close the browser.
+This step is important to prevent child processes persisting in the background after our program has finished.
+
+Now let's get to the tests!
+The idea is to control the browser in the same way we would expect a user to in each scenario we want to test.
+This means we use Puppeteer to type text into input forms, click buttons and search for particular elements on the page.
+In order to find those elements, we do need to make some adjustments in our React components, which we'll show later.
+Let's start at a higher level with a ``test``.
+
+.. literalinclude:: code/ui-before/index.test.tsx
+  :language: tsx
+  :start-after: // BEGIN_LOGIN_TEST
+  :end-before: // END_LOGIN_TEST
+
+We'll walk though this step by step.
+
+    - The ``test`` syntax is provided by Jest to indicate a new test running the function given as an argument (along with a description and time limit).
+    - ``getParty()`` gives us a new party name. Right now it is just a string unique to this set of tests, but in the future we will use the Party Management Service to allocate parties.
+    - ``newUiPage()`` is a helper function that uses the Puppeteer browser to open a new page (we use one page per party in these tests), navigate to the app URL and return a ``Page`` object.
+    - Next we ``login()`` using the new page and party name. This should take the user to the main screen. We'll show how the ``login()`` function does this shortly.
+    - We use the ``@daml/ledger`` library to check the ledger state. In this case, we want to ensure there is a single ``User`` contract created for the new party. Hence we create a new connection to the ``Ledger``, ``query()`` it and state what we ``expect`` of the result. When we run the tests, Jest will check these expectations and report any failures for us to fix.
+    - The test also simulates the new user logging out and then logging back in. We again check the state of the ledger and see that it's the same as before.
+    - Finally we must ``close()`` the browser page, which was opened in ``newUiPage()``, to avoid runaway Puppeteer processes after the tests finish.
+
+Selecting and interacting with UI elements
+==========================================
+
+We showed how to write a simple test at a high level, but haven't shown how to make individual actions in the app using Puppeteer.
+This was hidden in the ``login()`` and ``logout()`` functions.
+Let's see how ``login()`` is implemented.
+
+.. literalinclude:: code/ui-before/index.test.tsx
+  :language: tsx
+  :start-after: // BEGIN_LOGIN_FUNCTION
+  :end-before: // END_LOGIN_FUNCTION
+
+This looks somewhat mysterious at first.
+We first get some sort of handle to the username input element, and then use it to click into it and type the party name.
```

Comment:
Would write with more clarity and would avoid ambiguous terms like "somewhat mysterious", "some sort of a handle".